docs/rules/no-reference-type-as-default-prop.md

```diff
@@ -0,0 +1,68 @@
+# Prevent usage of referential-type variables as default param in functional component (react/no-reference-type-as-default-prop)
+
+Warns if in a functional component, a referential type value (such as array/object literal) is used as default prop, to prevent potential un-necessary re-renders, and performance regressions.
+
+## Rule Details
+
+Certain values (like arrays, objects, functions, etc) are compared by reference instead of by value. This means that, for example, whilst two empty arrays represent the same value - to the JS engine they are distinct and inequal as they represent references to different things in memory.
+
+When using object destructuring syntax you can set the default value for a given property if it does not exist. If you set the default value to one of the values that is compared by reference, it will mean that each time the destructure is executed the JS engine will create a new, non-equal value in the destructured variable.
+
+In the  context of a React functional component's props argument this means that each render the property has a new, distinct value. When this value is passed to a hook as a dependency or passed into a child component as a property react will see this as a new value - meaning that a hook will be re-evaluated, or a memoized component will rerender.
+
+This obviously destroys any performance benefits you get from memoisation. Additionally in certain circumstances this can cause infinite rerender loops, which can often be hard to debug.
```

Comment:
```suggestion
This obviously destroys any performance benefits you get from memoization. Additionally in certain circumstances this can cause infinite rerender loops, which can often be hard to debug.
```